play-json/shared/src/main/scala-3/play/api/libs/json/JsMacroImpl.scala

```diff
@@ -241,8 +241,12 @@ object JsMacroImpl { // TODO: debug
                   TypeRepr.of[String]
                 )
 
+                val termSym = tpr.termSymbol
+                val typeSym = tpr.typeSymbol
+                val isCase = termSym.flags.is(Flags.Case)
+
                 val tpeCaseName: Expr[String] = '{
-                  ${ config }.typeNaming(${ Expr(typeName(tpr.typeSymbol)) })
+                  ${ config }.typeNaming(${ Expr(typeName(if isCase then termSym else typeSym)) })
```

Comment:
By using the `termSymbol` we get the fully qualified name of the `enum` `case` rather than the name of the `enum` as our type discriminator.